### PR TITLE
Cleanup PaymentProcessor.doRefund API

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Minor cleanup docs, return values and exceptions. Also change from deprecated pass by reference on params to just returning a $params array.

Before
----------------------------------------
Less correct

After
----------------------------------------
More correct

Technical Details
----------------------------------------

Comments
----------------------------------------

### CODE REVIEW

File: api/v3/PaymentProcessor.php [hunk 1]
```diff
@@ -154,10 +155,15 @@ function _civicrm_api3_payment_processor_pay_spec(&$params) {
     'type' => CRM_Utils_Type::T_INT,
   ];
   $params['amount'] = [
-    'api.required' => TRUE,
```
Comment (eileenmcnaughton) [inline api/v3/PaymentProcessor.php]: I think TRUE is better  - ie  a boolean (1 will work but...)

Comment (mattwire) [reply api/v3/PaymentProcessor.php]: @eileenmcnaughton Fine by me :-)

File: api/v3/PaymentProcessor.php [hunk 2]
```diff
@@ -126,6 +126,7 @@ function _civicrm_api3_payment_processor_getlist_defaults(&$request) {
  * @throws \API_Exception
  */
 function civicrm_api3_payment_processor_pay($params) {
+  civicrm_api3_verify_mandatory($params, NULL, ['payment_processor_id', 'amount', 'contribution_id']);
```
Comment (eileenmcnaughton) [inline api/v3/PaymentProcessor.php]: We don't this call if they are all defined as required in _spec - which you have done below

File: CRM/Core/Payment.php [hunk 3]
```diff
@@ -1277,9 +1277,12 @@ public function doPayment(&$params, $component = 'contribute') {
    *
    * @param array $params
    *
+   * @return array
    * @throws \Civi\Payment\Exception\PaymentProcessorException
    */
-  public function doRefund(&$params) {}
+  public function doRefund(&$params) {
+    return [];
```
Comment (eileenmcnaughton) [inline CRM/Core/Payment.php]: what do we want to return? I'm keen to go with an exception is a failure & otherwise it's success because adding payment_status_id to the  array always felt hacky & we don't need to  return an array AND pass by ref

File: api/v3/PaymentProcessor.php [hunk 4]
```diff
@@ -158,6 +159,11 @@ function _civicrm_api3_payment_processor_pay_spec(&$params) {
     'title' => ts('Amount to pay'),
     'type' => CRM_Utils_Type::T_MONEY,
   ];
+  $params['contribution_id'] = [
```
Comment (eileenmcnaughton) [inline api/v3/PaymentProcessor.php]: Why is this required? The goal is to action a payment. Is it to ensure the info is always available to the processor & force people into the create-pending-contribution first flow?

File: api/v3/PaymentProcessor.php [hunk 5]
```diff
@@ -170,10 +176,11 @@ function _civicrm_api3_payment_processor_pay_spec(&$params) {
  * @throws CiviCRM_API3_Exception
  */
 function civicrm_api3_payment_processor_refund($params) {
+  /** @var \CRM_Core_Payment $processor */
   $processor = Civi\Payment\System::singleton()->getById($params['payment_processor_id']);
   $processor->setPaymentProcessor(civicrm_api3('PaymentProcessor', 'getsingle', array('id' => $params['payment_processor_id'])));
   if (!$processor->supportsRefund()) {
-    throw API_Exception('Payment Processor does not support refund');
+    throw new CiviCRM_API3_Exception('Payment Processor does not support refund');
```
Comment (eileenmcnaughton) [inline api/v3/PaymentProcessor.php]: Don't we throw API_Exception within apis & the wrapper throws the other?

**Comment by eileenmcnaughton** [comment]
@mattwire I added comments - I need more info as to why these changes are more correct  - shame we missed discussing these

File: api/v3/PaymentProcessor.php (hunk 4 above)
Comment (mattwire) [reply api/v3/PaymentProcessor.php]: See #15477

File: CRM/Core/Payment.php (hunk 3 above)
Comment (mattwire) [reply CRM/Core/Payment.php]: See follow-on in #15479.  I've kept them separate in terms of discussion that might be required :-)

File: api/v3/PaymentProcessor.php (hunk 2 above)
Comment (mattwire) [reply api/v3/PaymentProcessor.php]: Ok, removed.

File: api/v3/PaymentProcessor.php (hunk 5 above)
Comment (mattwire) [reply api/v3/PaymentProcessor.php]: It would appear so!

**Review by mattwire — COMMENTED** [review]
@eileenmcnaughton Thanks for review - I've updated per comments

File: api/v3/PaymentProcessor.php [hunk 6]
```diff
@@ -128,6 +128,11 @@ function _civicrm_api3_payment_processor_getlist_defaults(&$request) {
 function civicrm_api3_payment_processor_pay($params) {
   $processor = Civi\Payment\System::singleton()->getById($params['payment_processor_id']);
   $processor->setPaymentProcessor(civicrm_api3('PaymentProcessor', 'getsingle', ['id' => $params['payment_processor_id']]));
+
+  // Add some aliases that doPayment expects. These should go away once we've got getters for them in CiviCRM core.
+  // Historically doPayment has normally been passed 'contributionID'
+  $params['contributionID'] = $params['contribution_id'];
```
Comment (eileenmcnaughton) [inline api/v3/PaymentProcessor.php]: This is part of the pay api - I'm just gonna focus on that one until it's merged since there is a bit of a jumble of things happening here. I'll also look at the new docs

**Comment by eileenmcnaughton** [comment]
@mattwire I added https://lab.civicrm.org/dev/financial/issues/92 for the pass-by-params & merged the comments here https://github.com/civicrm/civicrm-core/pull/15607 - as a reviewer's commit)
